File: problems/768_max_chunks_to_make_sorted_ii/solution.py
from typing import List


class Solution:
    def maxChunksToSorted(self, arr: List[int]) -> int:
        stack = []
        for num in arr:
            mx = num
            while stack and stack[-1] > num:
                mx = max(mx, stack.pop())
            stack.append(mx)
        return len(stack)

File: problems/768_max_chunks_to_make_sorted_ii/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int maxChunksToSorted(vector<int>& arr) {
        stack<int> stk;
        for (int num : arr) {
            int mx = num;
            while (!stk.empty() && stk.top() > num) {
                mx = max(mx, stk.top()); stk.pop();
            }
            stk.push(mx);
        }
        return stk.size();
    }
};
